src/ansi-c/library/stdlib.c

```diff
@@ -60,27 +60,40 @@ inline void abort(void)
 /* FUNCTION: calloc */
 
 #undef calloc
-#undef malloc
-
-inline void *malloc(__CPROVER_size_t malloc_size);
 
 inline void *calloc(__CPROVER_size_t nmemb, __CPROVER_size_t size)
 {
+  // realistically, calloc may return NULL,
+  // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
-  void *res;
-  res=malloc(nmemb*size);
+  // ensure that all bytes in the allocated memory can be addressed
+  // using our object:offset encoding as specified in
+  // flattening/pointer_logic.h; also avoid sign-extension issues
+  // for 32-bit systems that yields a maximum allocation of 2^23-1,
+  // i.e., just under 8MB
+  __CPROVER_assume(nmemb*size<(1ULL<<((sizeof(char*)-1)*8-1)));
+  void *malloc_res;
+  malloc_res=__CPROVER_allocate(nmemb*size, 1);
+
+  // make sure it's not recorded as deallocated
+  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+
+  // record the object size for non-determistic bounds checking
+  __CPROVER_bool record_malloc;
+  __CPROVER_malloc_object=record_malloc?malloc_res:__CPROVER_malloc_object;
+  __CPROVER_malloc_size=record_malloc?nmemb*size:__CPROVER_malloc_size;
+  __CPROVER_malloc_is_new_array=record_malloc?0:__CPROVER_malloc_is_new_array;
+
+  // detect memory leaks
+  __CPROVER_bool record_may_leak;
+  __CPROVER_memory_leak=record_may_leak?malloc_res:__CPROVER_memory_leak;
```

Comment:
Do we have an explicit get-nondet-bool so this looks less like a mistake to the casual reader? Otherwise suggest commenting to this effect.